Bump gcc to gcc10 and binutils to 2.34

luc65r opened the pull request:
## Motivation for this change
I wanted to have a try at updating binutils (I took the work of @lovesegfault #86954), and as all packages needed to be recompiled, I tried to make gcc10 default.

## Things done

- [x] Fix `ld: multiple definition of` (made gcc10 use `-fcommon` by default)
- Fix packages to work with `-fnocommon`
   - [x] gdbm
- [x] Fix python's `_findLib_ld`
- [x] Fix libomxil-bellagio build
- [x] Fix game-music-emu build (updated it)
- [x] Fix rEFInd build (TODO: build it with gcc10)
- [x] Fix dhcp build
- [x] Fix pythonPackages.cffi (TODO: make tests pass)
- [x] Fix pythonPackages.numpy (`ImportError: liblapack.so.3: ELF load command address/offset not properly aligned`)
- [x] Fix systemd

### Make clean fixes for

- [x] rEFInd
- [x] pythonPackages.cffi
- [x] blas and lapack
- [x] libopus
- [x] qt512
- [x] ncmpcpp
- [x] libreoffice
- [x] range-v3
- [x] zimg

___

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

Commit:
use gcc10 instead of gcc9

Commit:
gcc: make -fcommon default on gcc10

Commit:
libomxil-bellagio: fix build with gcc10

Commit:
game-music-emu: 0.6.1 -> 0.6.3

Commit:
dhcp: fix build with gcc10

FRidh commented:
We were at binutils 2.34 one or twice before but it got reverted.

luc65r commented:
@FRidh We need to make sure that most of the things work before merging.
Is it fine to do multiple things in one PR? I don't want to compile my system 2 times, it takes already long enough.

FRidh commented:
Please look at the past work that has been done on 2.34. This involves more work than you probably imagine right now.

luc65r commented:
Yes, I saw that, I have a lot of free time (not a lot of skills), and I hope people will help.

lovesegfault commented:
While I appreciate the effort to try and bump binutils, I have a few comments:
1. It's not great that you chose to rewrite the previous patches and submit them without any authorship of the work that was done before to try and get this to land.
2. Not sure the right approach is to go about reverting upstream commits. I recommend you go ask their mailing list so that, if not now in the future, we can avoid needing to maintain downstream patches.
3. If you need someone to test things that requires lots of rebuilds ping me on IRC.

Commit:
Revert "Revert "Merge pull request #86954 from lovesegfault/binutils-2.34""

This reverts commit 24c96b92595e18e73645715427bba32adbaf7677.

Commit:
Revert "Revert "Revert "Revert "libbfd: fix build""""

This reverts commit 5c2b59d63afe75d6735790ca9232a1445091bf53.

luc65r commented:
>     1. It's not great that you chose to rewrite the previous patches and submit them without any authorship of the work that was done before to try and get this to land.
Sorry, I was stupid.
>     2. Not sure the right approach is to go about reverting upstream commits. I recommend you go ask their mailing list so that, if not now in the future, we can avoid needing to maintain downstream patches.
You are right, should I try to patch python instead?

I'm a bit stuck with numpy (who would have thought...), I can't get it to output useful error messages.

luc65r commented:
> That's not really a fix.

It isn't, but I do not have the technical abilities to make a proper fix, I don't understand the install phases of lapack and blas.
It enables me to try fixing other packages which depends on numpy.

Commit:
perlPackages.ExtUtilsPkgConfig: disable checks

Commit:
pythonPackages.coveralls: delete restrictions for coverage version

luc65r commented:
I'm really stuck on systemd...
> This involves more work than you probably imagine right now.

I never thought systemd could brake so badly xD

It fails to start a lot a services like `journald`, `logind` or `udevd` with error code 127.
It seems related to binutils, as it has the same behavior compiled with `gcc9` and `gcc10`.

luc65r commented:
Everything seems to work now!

Is calling `execstack` on `libgcrypt.so.20` safe?

There is still one big thing to do, namely make `pythonPackages.numpy` use `lapack` and `blas`, so if you guys could help me, it would be great.

lovesegfault commented:
@luc65r Are you still working on this?

luc65r commented:
@lovesegfault Not lately. I think it can be merged, because it won't break a lot of things. There is just `numpy` with `blas` and `lapack` which need to be fixed, but I can't do it myself, so someone else can make a PR for that.

FRidh commented:
No, it can't be merged yet because the issue for which the binutils bump got reverted is not fixed yet (https://github.com/NixOS/nixpkgs/commit/24c96b92595e18e73645715427bba32adbaf7677).

luc65r commented:
#98f335a seems to have fixed numpy

pbogdan commented:
Probably out of scope for now anyway but just a note that binutils 2.35 was branched off last weekend, with a release expected in 2 weeks time.

Commit:
syslinux: build with gcc9

build fails with "undefined reference to `strlen'" on gcc10

Commit:
vsftpd: fix build with gcc10

Commit:
searx: fix build

luc65r commented:
Now can it be merged? The only problems I can see are the disabled checks on some packages, but I checked myself that they work.

lovesegfault commented:
@FRidh I wonder whether we could set up a separate hydra job to eval this branch. Just to avoid it breaking `staging` all the time (as happened in the previous binutils bump attempt)

Ma27 commented:
@luc65r you probably want to rebase onto latest staging.
cc @vcunat @FRidh re/merge.

jonringer commented:
is this on the latest staging?

jonringer commented:
also, it might be easier to do the gcc bump or the binutils bump, but both complicates regressions significantly

tobim commented:
> also, it might be easier to do the gcc bump or the binutils bump, but both complicates regressions significantly

Bumping gcc before might actually break more things than updating both at the same time. The best option would be to create a hydra job for the combined update as @lovesegfault suggested. If it breaks badly the next try would be to bump binutils only.

FRidh commented:
This is too big of a change before a release. These two bumps always come with more regressions and are best done a month or so after a release.

jonringer commented:
> The best option would be to create a hydra job for the combined update as @lovesegfault suggested. If it breaks badly the next try would be to bump binutils only.

didn't realize it didn't

vcunat commented:
IIRC we repeatedly had quite some problems with binutils updates in the past several years, so yes... I most likely wouldn't aim for 20.09 #95475.  But I can't see anything preventing merge to staging soon after the fork-off in a few weeks, provided that a separate jobset looks good.

mkg20001 commented:
@luc65r @vcunat The branch-off for 20.09 happend some days ago. What's with this PR now?

FRidh commented:
The first weeks after branch-off Hydra tends to be overloaded, so running such a job is not a very good idea then.

luc65r commented:
@mkg20001 It works since July 6, I built my whole system with it. I haven't tested Gnome/KDE stuff though.
I don't know if anyone went through the changes, it might be a good idea as it is kind of important security-wise.
Stuff could be cleaned up, but I don't really want to do it until it has a chance to be merged.

mkg20001 commented:
@FRidh So I can merge it in a week or so?

vcunat commented:
No, I think we should first create a separate jobset to estimate how big the overall breakage is.  I hope the build farm will free up a little within a couple weeks.

vcunat commented:
IIRC some of the binutils builds broke lots of things and weren't easy to integrate; I don't think I can predict it well.

vcunat commented:
To be clear, world-rebuiding changes do get merged almost every week to staging, so saving machine time probably won't be sufficient motivation to put another fresh version in that hasn't been really tested here by anyone yet.

luc65r commented:
Testing binutils-2.35 right now, the bump seems pretty straight forward.

lovesegfault commented:
I really don't see a way to merge this with any level of confidence unless we make a hydra jobset for this PR and evaluate it by itself. We can't expect @luc65r to build the world by himself.

vcunat commented:
`x86_64-linux` is quite freed up on Hydra ATM, so I started with a small NixOS jobset: https://hydra.nixos.org/eval/1614356

_I'm sorry about the mess shown here.  I merged `master` without realizing many of its commits haven't made it to the target `staging` branch yet._

vcunat commented:
Most of the jobset failed, so there's certainly still work to do...

tobim commented:
Quite a lot of these seem to be due to `fn-common` being default now: https://wiki.gentoo.org/wiki/Gcc_10_porting_notes/fno_common

LFS has fixes for may of those: [cpio](http://www.linuxfromscratch.org/blfs/view/svn/general/cpio.html), [sharutils](http://www.linuxfromscratch.org/blfs/view/svn/general/sharutils.html).

vcunat commented:
Some distros have been using gcc10 for a long time (e.g. Fedora), so I'm confident that fixes can be found for those... but it may still be substantial work on our side to pull it all in.

luc65r commented:
@vcunat I don't really understand why you did that.
It would have been merged to staging, and then to master.
Could you revert that?

> Quite a lot of these seem to be due to fno-common being default now
I started taking patches, but I realized it would be a lot of work, so I added `-fcommon` to CFLAGS (can't really find it with all this mess).

Before master was merged here, all the things that I could test worked fine and I successfully build my entire system with this pr.

vcunat commented:
The GitHub UI problem got easily fixed by the retargeting dance seen above.  I don't think undoing the merge of master would help anything, and master gets merged to staging anyway regularly (often indirectly through staging-next).  If such a merge breaks some package, it's surely better to find it _before_ merging this PR to staging (but I don't think it should be common).

vcunat commented:
Adding `-fcommon` to _individual_ packages certainly seems OK in cases there's no other easy fix.

Mic92 commented:
@vcunat should we cherry-pick fixes or do you want to merge it as a whole?

Commit:
Remove trailing whitespaces and fix indentation

luc65r commented:
The [Hydra jobset](https://hydra.nixos.org/jobset/nixos/binutils-2.34) fails evaluating with this error:
```
hydra-eval-jobs returned exit code 1:
error: "error: --- ThrownError --- hydra-eval-jobs\nThe option `virtualisation.qemu' does not exist. Definition values:\n- In `/nix/store/ls43vcjhan3yfg1g5cpr6xqa5cqqx6sj-source/nixos/modules/testing/test-instrumentation.nix':\n    {\n      consoles = [ ];\n      package = {\n        _type = \"override\";\n        content = <derivation /nix/store/ggw9cqx85hy1nmzbr1nsmbylsfvj7b3j-qemu-host-cpu-only-for-vm-tests-5.1.0.drv>;\n    ..."
error: [json.exception.type_error.302] type must be string, but is null
```
What can I do?

vcunat commented:
That's a problem we had on `master` at some point.  Merging latest `master` (or rebasing) should solve that.

vcunat commented:
I merged master to staging, in case that helps.

vcunat commented:
@Mic92: I don't mind either.  If you need some of these fixes, certainly feel free to cherry-pick them.

If we'd expect this branch to stabilize and merge very fast, I expect it's less work to merge it only together (except for particular commits if you need them for something).  If it could take long time, it would make sense for the fixes to be merged ahead.

luc65r commented:
All the jobs in [the jobset](https://hydra.nixos.org/jobset/nixos/binutils-2.34) succeeded, but it was only `nixos/release-small.nix`. We could maybe try with `pkgs/top-level/release.nix` now.

vcunat commented:
Right, I switched it to [all x86_64 packages](https://hydra.nixos.org/eval/1622768).  The other two platforms are quite busy at the moment, so this seems relatively convenient.

luc65r commented:
I guess 91% of job success rate isn't bad considering it's based on staging

luc65r commented:
Rebased against staging, systemd fix is no longer needed

Commit:
Revert commits on binutils for OpenRISC 1000

This reverts commits
    ebd89f6de1894f3c3d22e4bb8cb879be7e9556b3
    c6f6db77ac53e84006353396283d271cb1b22460
    257fb62387f1324a9cb75465dceb3d84eefc2b4d
    dbe556af9cb9d00c6c50d76a1c1abade5455ab2f

FRidh commented on pkgs/data/fonts/noto-fonts/tools.nix:
```diff
@@ -70,6 +70,8 @@ buildPythonPackage rec {
     cp -r third_party $out
   '';
 
+  dontUsePythonRecompileBytecode = true;
```
This can be removed, see https://github.com/NixOS/nixpkgs/pull/103089.

FRidh commented on pkgs/development/libraries/liblangtag/default.nix:
```diff
@@ -39,12 +39,12 @@ stdenv.mkDerivation rec {
   nativeBuildInputs = [ autoreconfHook gtk-doc gettext pkgconfig ];
 
-  meta = {
+  meta = with stdenv.lib; {
```
please avoid making these kind of style changes, because they're typically personal.

FRidh commented on pkgs/development/libraries/libopus/default.nix:
```diff
@@ -18,7 +18,7 @@ stdenv.mkDerivation {
                 ++ stdenv.lib.optional withCustomModes "--enable-custom-modes";
 
-  doCheck = true;
+  doCheck = !stdenv.isi686;
```
Motivation?

luc65r replied:
I have no idea...

FRidh commented on pkgs/development/python-modules/cffi/default.nix:
```diff
@@ -32,7 +32,7 @@ if isPyPy then null else buildPythonPackage rec {
     "-Wno-unused-command-line-argument -Wno-unreachable-code";
 
-  doCheck = !stdenv.hostPlatform.isMusl && !stdenv.isDarwin; # TODO: Investigate
+  doCheck = false; #!stdenv.hostPlatform.isMusl && !stdenv.isDarwin; # TODO: Investigate
```
This will need to be fixed before we merge. I can have a look at that.

luc65r replied:
A lot of `AssertionError: assert 'linux' == 'win32'` and `OSError: ctypes.util.find_library() did not manage to locate a library called 'm'`

FRidh replied:
What is the status on this part?

luc65r replied:
The errors sadly didn't magically disappear... I will try to make the tests pass tomorrow.

luc65r replied:
I fixed the tests on 3.7 and 3.8 by backporting a fix of ctypes.util.find_library.

FRidh commented on pkgs/development/python-modules/freezegun/default.nix:
```diff
@@ -23,6 +23,8 @@ buildPythonPackage rec {
   # contains python3 specific code
   doCheck = !isPy27;
 
+  dontUsePythonRecompileBytecode = true;
```
This can be removed, see #103089.

FRidh commented on pkgs/development/python-modules/google_apputils/default.nix:
```diff
@@ -30,6 +30,8 @@ buildPythonPackage rec {
   # ERROR:root:Trying to access flag test_tmpdir before flags were parsed.
   doCheck = false;
 
+  dontUsePythonRecompileBytecode = true;
```
This can be removed, see #103089.

FRidh commented on pkgs/development/python-modules/jinja2/default.nix:
```diff
@@ -27,6 +27,8 @@ buildPythonPackage rec {
     pytest -v tests
   '';
 
+  dontUsePythonRecompileBytecode = true;
```
This can be removed, see #103089.

FRidh commented on pkgs/development/python-modules/mox/default.nix:
```diff
@@ -15,6 +15,8 @@ buildPythonPackage rec {
   # error: invalid command 'test'
   doCheck = false;
 
+  dontUsePythonRecompileBytecode = true;
```
This can be removed, see #103089.

FRidh commented on pkgs/development/python-modules/proboscis/default.nix:
```diff
@@ -12,6 +12,8 @@ buildPythonPackage rec {
   propagatedBuildInputs = [ nose ];
   doCheck = false;
 
+  dontUsePythonRecompileBytecode = true;
```
This can be removed, see #103089.

FRidh commented on pkgs/development/python-modules/soco/default.nix:
```diff
@@ -21,6 +21,8 @@ buildPythonPackage rec {
     sphinx_rtd_theme
   ];
 
+  doCheck = false;
```
Motivation?

FRidh commented on pkgs/development/python-modules/testtools/default.nix:
```diff
@@ -30,6 +30,8 @@ buildPythonPackage rec {
     substituteInPlace requirements.txt --replace "fixtures>=1.3.0" ""
   '';
 
+  dontUsePythonRecompileBytecode = true;
```
This can be removed, see #103089.

FRidh commented on pkgs/tools/typesetting/tex/texlive/bin.nix:
```diff
@@ -341,6 +341,8 @@ pygmentex = python2Packages.buildPythonApplication rec {
     runHook postInstall
   '';
 
+  dontUsePythonRecompileBytecode = true;
```
This can be removed, see #103089.

FRidh reviewed (changes requested):
I made some minor comments. These can be addressed for when we merge this into `staging-next` or `master`. Other than that this looks good!

A quick look at the Hydra jobset shows that Qt 5.14, specifically qtlocation-5.14.2, needs fixing. Mutter as well.

jonringer commented on pkgs/development/libraries/gdbm/default.nix:
```diff
@@ -9,6 +9,11 @@ stdenv.mkDerivation rec {
     sha256 = "1p4ibds6z3ccy65lkmd6lm7js0kwifvl53r0fd759fjxgr917rl6";
   };
 
+  patches = [
+    # Taken from https://bugs.gentoo.org/705898
+    ./gdbm-1.18.1-gcc10.patch
+  ];
+
```
```
Automatic merge went well; stopped before committing as requested
error: attribute 'patches' at /home/jon/.cache/nixpkgs-review/pr-89793/nixpkgs/pkgs/development/libraries/gdbm/default.nix:19:3 already defined at /home/jon/.cache/nixpkgs-review/pr-89793/nixpkgs/pkgs/development/libraries/gdbm/default.nix:12:3
(use '--show-trace' to show detailed location information)
```

jonringer commented:
should we rebase this on the latest staging, and push it to a `NixOS/nixpkgs` branch with a hydra jobset?

vcunat commented:
This has a jobset already: https://hydra.nixos.org/jobset/nixos/binutils-2.34

Commit:
liblangtag: 0.6.1 -> 0.6.3

luc65r commented:
> A quick look at the Hydra jobset shows that Qt 5.14, specifically qtlocation-5.14.2, needs fixing. Mutter as well.

I think a lot of job failure is due to the branch being based on staging.
Most of the jobs that fail here fail on staging too. The thing is I wont compile all staging to compare the jobs that fail.

vcunat commented:
For this reason I think it would be more practical to keep rebasing such branches on `staging-next` or even `master` (even though still being filed for merging to `staging`).

luc65r commented:
I have some free time now, and I would really want to get this merged soon, I'm tired of resolving merge conflicts.
What is left to do?
Should I rebase it against master even though it will fuck up Github's interface?

TredwellGit commented:
I can help test this, but it should be updated to binutils 2.35.1.

jonringer commented on pkgs/tools/networking/dhcp/default.nix:
```diff
@@ -44,6 +44,7 @@ stdenv.mkDerivation rec {
     "-Wno-error=format-truncation"
     "-Wno-error=stringop-truncation"
     "-Wno-error=format-overflow"
+    "-Wno-error=stringop-overflow=8"
```
not your fault, but:
```
 pkgs/tools/networking/dhcp/default.nix:
	71: Wrong indent style found (tabs instead of spaces)
```

jonringer reviewed (commented):
pkgs/tools/networking/dhcp/default.nix:
	71: Wrong indent style found (tabs instead of spaces)

r-burns commented:
I would suggest adding a comment to each `Wno-error`, `gcc9Stdenv`, etc fix pointing to the relevant upstream issue tracker. Ideally we will remove these one-off fixes as they are resolved upstream rather than carrying them around indefinitely.

r-burns commented on pkgs/top-level/all-packages.nix:
```diff
@@ -22923,7 +22928,9 @@ in
   ncmpc = callPackage ../applications/audio/ncmpc { };
 
-  ncmpcpp = callPackage ../applications/audio/ncmpcpp { };
+  ncmpcpp = callPackage ../applications/audio/ncmpcpp {
+    stdenv = gcc9Stdenv;
```
https://github.com/ncmpcpp/ncmpcpp/commit/cc6fbddb3798346e20d4a694ff107b5976ab060e

r-burns commented on pkgs/top-level/all-packages.nix:
```diff
@@ -15824,7 +15824,9 @@ in
   raft-canonical = callPackage ../development/libraries/raft-canonical { };
 
-  range-v3 = callPackage ../development/libraries/range-v3 {};
+  range-v3 = callPackage ../development/libraries/range-v3 {
+    stdenv = gcc9Stdenv;
```
https://github.com/ericniebler/range-v3/commit/a91f0e1be27a31c446452a753001d4518ef83a6b

r-burns commented on pkgs/top-level/all-packages.nix:
```diff
@@ -19255,7 +19257,9 @@ in
   sysklogd = callPackage ../os-specific/linux/sysklogd { };
 
-  syslinux = callPackage ../os-specific/linux/syslinux { };
+  syslinux = callPackage ../os-specific/linux/syslinux {
+    stdenv = gcc9Stdenv;
```
https://bugs.gentoo.org/705730
https://gitweb.gentoo.org/repo/gentoo.git/commit/?id=8284530f4bd86a38b9867f847bd844f5d8d08e87

luc65r replied:
It fails with ``undefined reference to `strlen'``
https://wiki.syslinux.org/wiki/index.php?title=Building

r-burns commented on pkgs/tools/archivers/p7zip/default.nix:
```diff
@@ -42,7 +42,8 @@ stdenv.mkDerivation rec {
   setupHook = ./setup-hook.sh;
 
-  NIX_CFLAGS_COMPILE = stdenv.lib.optionalString stdenv.cc.isClang "-Wno-error=c++11-narrowing";
+  NIX_CFLAGS_COMPILE = stdenv.lib.optionalString stdenv.cc.isClang "-Wno-error=c++11-narrowing"
+    + "-Wno-error=narrowing";
```
https://sourceforge.net/p/p7zip/bugs/226/

(Also, this will break on clang, the flags need to be space-separated)

jonringer commented:
luc65r, I would also recommend using a different branch name than something that exists on upstream. A lot of tools will try to checkout with your branch name, and they will fail if there's already a branch of that name

luc65r commented:
> I would also recommend using a different branch name than something that exists on upstream

I don't think I can without making a new pr @jonringer 

> it should be updated to binutils 2.35.1

should it really?
I tested updating it, it's pretty straightforward, but we could maybe stay with 2.34 and update it when this pr is merged.

Commit:
cpython: fix ctypes.util.find_library

Commit:
refind: fix build on gcc10

Commit:
p7zip: fix build on gcc10

Commit:
range-v3: fix build on gcc10

lovesegfault commented:
I agree with @luc65r here. I'd much rather do the bump to 2.34 first, and then go to 2.35.1.

This is difficult enough as it is.

jonringer commented:
agreed, should mitigate risk when there's already enough

dont want to bite off more than you can chew. And this is already a large bite

TredwellGit commented:
Tentative list of fixes:
```
--- a/pkgs/top-level/all-packages.nix
+++ b/pkgs/top-level/all-packages.nix
@@ -10482,7 +10482,11 @@ in

   groovy = callPackage ../development/interpreters/groovy { };

-  guile_1_8 = callPackage ../development/interpreters/guile/1.8.nix { };
+  guile_1_8 = callPackage ../development/interpreters/guile/1.8.nix {
+    buildPackages = buildPackages // {
+      stdenv = gcc9Stdenv;
+    };
+  };

   # Needed for autogen
   guile_2_0 = callPackage ../development/interpreters/guile/2.0.nix { };
```

Commit:
libreoffice-still: build with gcc9

Fails in multiple ways when building with gcc10.
Moving to a newer version may solve this.

luc65r commented:
> ```
> -  guile_1_8 = callPackage ../development/interpreters/guile/1.8.nix { };
> +  guile_1_8 = callPackage ../development/interpreters/guile/1.8.nix {
> +    buildPackages = buildPackages // {
> +      stdenv = gcc9Stdenv;
> +    };
> +  };
> ```

@TredwellGit `guile_1_8` builds fine with gcc10

TredwellGit commented:
I am testing this with https://github.com/NixOS/nixpkgs/pull/104091 and apparently guile_1_8 with gcc10 does not build as a PIE, but requires other security flags enabled to build which is rather weird. With gcc9 it builds as a PIE. Ignore what I posted; I will fix this in https://github.com/NixOS/nixpkgs/pull/104091.

TredwellGit commented:
nixUnstable does not build for me.

TredwellGit commented:
pkgsi686Linux.libopus does not build either.

Commit:
libopus: disable tests on i686

test_unit_LPC_inv_pred_gain fails

Commit:
nixUnstable: fix build on gcc10

fabianhjr commented:
There is a comment for binutils that should be included:

```
pkgs/misc/emulators/wine/base.nix:  # TODO: binutils 2.34 contains a fix for this bug, re-enable stripping once available.
pkgs/misc/emulators/wine/base.nix-  dontStrip = true;
pkgs/misc/emulators/wine/base.nix-
```

Also tried this branch on my computer with systemd/gdm/gnome and seems like this branch cannot boot. It cannot start journal and disks among others.

fabianhjr commented:
`pkgs/applications/audio/ncmpcpp` was upgraded on NixOS master and staging to 0.9.1 and that is the only conflict with staging.

Commit:
wine: re-enable stripping

luc65r commented:
> There is a comment for binutils that should be included:

Thanks! I wouldn't have noticed.

> It cannot start systemd-journald among other systemd services.

Pfff... Not again...

luc65r commented:
The reason systemd services cannot start is `libgcrypt.so.20: cannot enable executable stack as shared object requires: Operation not permitted`
I made a fix (by executing `execstack`), but the commit has disappeared somehow (I don't know how I did that...).
Investigating a bit further, I found out that libgcrypt isn't compiled with `--noexecstack` because the configure script has decided gcc10 doesn't support `--noexecstack`:
```
configure:17628: checking whether assembler supports --noexecstack option
configure:17638: gcc -g -O2 -fvisibility=hidden -fno-delete-null-pointer-checks -Wall
                     -S -o conftest.s conftest.c >/dev/null
configure:17641: $? = 0
configure:17646: gcc -g -O2  -Wa,--noexecstack
                       -c -o conftest.o conftest.s >/dev/null
conftest.s: Assembler messages:
conftest.s: Fatal error: duplicate .debug_line sections
configure:17649: $? = 1
configure:17658: result: no
```
What should I do? Patch the configure script? Use `execstack`?

Commit:
libgcrypt: force noexecstack

fabianhjr commented:
I would prefer patching the configure of libgcrypt (which looking at the commit is what you did)

Building an image for testing, thangs luc65r :3

fabianhjr commented:
Good news, after ~4 hours of compile time it does boot and seems to work great:

```
 ~  cat /proc/version                                                                                   Sat 26 Dec 2020 10:03:19 PM UTC
Linux version 5.10.2 (nixbld@localhost) (gcc (GCC) 10.2.0, GNU ld (GNU Binutils) 2.34) #1-NixOS SMP Mon Dec 21 12:30:08 UTC 2020
```

ajs124 commented on pkgs/top-level/all-packages.nix:
```diff
@@ -19272,7 +19271,9 @@ in
 
-  syslinux = callPackage ../os-specific/linux/syslinux { };
+  syslinux = callPackage ../os-specific/linux/syslinux {
+    stdenv = gcc9Stdenv; # Fails with "undefined reference to `strlen'" on gcc10
```
OpenSUSE seems to have a patch for this one: https://build.opensuse.org/package/view_file/openSUSE:Factory/syslinux/syslinux-4.04-gcc10.diff?expand=1

luc65r replied:
Thanks!

Mic92 commented:
@luc65r thanks for seeing through this.

Commit:
syslinux: fix build on gcc10

Commit:
mp4v2: fix build on gcc10

FRidh commented:
I've merged the parts of staging that this was based on into staging-next (92cc19490e7fe141257dbb840ad518c7156f0efb), and then I merged this PR into staging-next as well (f3139e90caeeabd5953b4c97e1c04cb430e9e06f). That will save us a rebuild.

fabianhjr commented:
Thanks @FRidh, @luc65r, et al. for working on this. :3

Will take a look at next-staging tomorrow and help stabilize it.

FRidh commented:
Several aarch64 builds fail with `bin/ld: final link failed: bad value`
```
ar: `u' modifier ignored since `D' is the default (see `U')
  CXXLD    libprotoc.la
libtool: warning: '/nix/store/spps8qis3wmbr5051g64l1lzv82d7zsc-gcc-10.2.0-lib/lib/libstdc++.la' seems to be moved
ar: `u' modifier ignored since `D' is the default (see `U')
  CXXLD    protoc
/nix/store/d3a24znms2blwv2fvqcrmrhwqq08f425-binutils-2.34/bin/ld: .libs/protoc: hidden symbol `__aarch64_cas8_rel' in /nix/store/9zvbxwpzfp9k6xl9px67y0l2m8xy2d1w-gcc-10.2.0/lib/gcc/aarch64-unknown-linux-gnu/10.2.0/libgcc.a(cas_8_3.o) is referenced by DSO
/nix/store/d3a24znms2blwv2fvqcrmrhwqq08f425-binutils-2.34/bin/ld: final link failed: bad value
collect2: error: ld returned 1 exit status
make[2]: *** [Makefile:3892: protoc] Error 1
make[2]: Leaving directory '/build/source/src'
make[1]: *** [Makefile:1874: all-recursive] Error 1
make[1]: Leaving directory '/build/source'
make: *** [Makefile:1781: all] Error 2
builder for '/nix/store/cl5fmbdqxnkvmfw21s5wwz2lrqpc72lq-protobuf-3.14.0.drv' failed with exit code 2
```

luc65r commented:
I didn't test on aarch64 and darwin at all, we might have to force gcc9 on them for now.

vcunat commented:
Darwin uses clang as default.

flokli commented:
I can reproduce the build error with `nix-build -A protobuf` on a aarch64-linux box.

I can't reproduce it when cross-compiling from x86_64.

MetaDark commented:
Thanks everyone for working on this!

I noticed two small issues building my system with 4008fd38417:
- `python-prctl` fails on `test_mpx` even with #106875 cherry-picked:
  It looks like mpx support should have been removed in GCC 9.1, so it's strange we're now only seeing this issue with GCC 10: https://www.phoronix.com/scan.php?page=news_item&px=MPX-Removed-From-GCC9
- `ccls` fails to build since llvmPackages_8 fails to build:
  [ccls-failure.log](https://github.com/NixOS/nixpkgs/files/5747981/ccls-failure.log)


I noticed that stripping was re-enabled on wine, so I will also confirm if binutils 2.34 fixes the original issue.

fabianhjr commented:
Just as a heads up, it seems like the staging-next hydra job runs too often and currently has 2 runs building the world: https://hydra.nixos.org/jobset/nixpkgs/staging-next

FRidh commented:
cc @NixOS/exotic-platform-maintainers for assistance in fixing up aarch64. See https://github.com/NixOS/nixpkgs/pull/107783.

FRidh commented:
suggested aarch64 fix https://bugzilla.redhat.com/show_bug.cgi?id=1830472#c1

> On aarch64 -moutline-atomics has been turned on by default, and those symbols are solely in libgcc.a, not in libgcc_s.so.*.
So, libgcc_s.so linker script is needed instead of libgcc_s.so symlink like on some other arches, gcc-10.0.1-0.14.fc33 with a fix is building in koji.

vcunat commented:
Sounds it could be "the same problem".  Their patch: https://src.fedoraproject.org/rpms/gcc/c/95507e8b68

FRidh commented:
They remove `libgcc_s.so` and then replace it with
```
GROUP ( /%{_lib}/libgcc_s.so.1 libgcc.a )' > $FULLPATH/libgcc_s.so
```
If I am correct `libgcc_s.so` is actually a statically linked file despite the .so, and this line here merges the two libraries into a new one, which is I think also what they meant with the message I quoted.

We have  `libgcc_s.so` in the glibc store path, not in the gcc path. It is pulled in by the wrapper in the file `/nix/store/y78inja64xaaq6bx33nxn2h1lgw0jswk-gcc-wrapper-10.2.0/nix-support/libc-crt1-cflags` which contains
```
-B/nix/store/rqrklqsvw4ydpcg5kdcvn506fhcbqxk2-glibc-2.32-10/lib/
```
This is defined in `pkgs/build-support/cc-wrapper/default.nix`
```
echo "-B${libc_lib}${libc.libdir or "/lib/"}" >> $out/nix-support/libc-crt1-cflags
```
Note also the comment
```
# The "-B${libc_lib}/lib/" flag is a quick hack
```

luc65r commented:
The tests for `pkgsi686Linux.libopus` pass with https://github.com/xnorpx/opus/commit/6e2209d3dcdd4c4cf3b38f9357ce829d932c8b48.
